libselinux: Mount procfs before checking /proc/filesystems

In the case where the SELinux security module is not loaded in the
kernel and it's early enough in the boot process that /proc has not yet
been mounted, selinuxfs_exists() will incorrectly return 1.

Mount procfs before opening /proc/filesystems, then unmount it if we
mounted it. Same approach selinux_init_load_policy() uses for
/proc/cmdline.

Signed-off-by: Ben Shelton <ben.shelton@ni.com>

Upstream-Status: Inappropriate [OpenXT early-boot /proc not mounted]

--- a/src/init.c	2026-09-10 10:56:49.355753415 -0400
+++ b/src/init.c	2026-09-10 10:56:49.355753415 -0400
@@ -11,6 +11,7 @@
 #include <sys/vfs.h>
 #include <stdint.h>
 #include <limits.h>
+#include <sys/mount.h>
 
 #include "policy.h"
 #include "selinux_internal.h"
@@ -54,15 +55,20 @@
 
 int selinuxfs_exists(void)
 {
-	int exists = 0;
+	int exists = 0, mnt_rc = 0;
 	FILE *fp = NULL;
 	char *buf = NULL;
 	size_t len;
 	ssize_t num;
 
+	mnt_rc = mount("proc", "/proc", "proc", 0, 0);
+
 	fp = fopen("/proc/filesystems", "re");
-	if (!fp)
-		return 1; /* Fail as if it exists */
+	if (!fp) {
+		exists = 1; /* Fail as if it exists */
+		goto out;
+	}
+
 	__fsetlocking(fp, FSETLOCKING_BYCALLER);
 
 	num = getline(&buf, &len, fp);
@@ -76,6 +82,14 @@
 
 	free(buf);
 	fclose(fp);
+
+out:
+#ifndef MNT_DETACH
+#define MNT_DETACH 2
+#endif
+	if (mnt_rc == 0)
+		umount2("/proc", MNT_DETACH);
+
 	return exists;
 }
 
